data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/RemotePeerForwarder.java

```diff
@@ -141,8 +141,17 @@ private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(
             final Event event = record.getData();
 
             final List<String> identificationKeyValues = new LinkedList<>();
+            boolean identificationKeysNotFound = false;
             for (final String identificationKey : identificationKeys) {
-                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
+                Object identificationKeyValue = event.get(identificationKey, Object.class);
+                if (identificationKeyValue == null) {
+                    identificationKeysNotFound = true;
+                    break;
+                }
+                identificationKeyValues.add(identificationKeyValue.toString());
```

Comment:
I think the behavior we want if an event does not have all the identification keys would be to just not add it to the values list. For example, let's say we have identification_keys `client`, `server`, `verb`, then the following event 

```
{
     "client": "1234",
     "server": "5678"
}
```

is still valid and would map to the identification_keys hash of 

```
{
     "client": "1234",
     "server": "5678",
     "verb": null
}
```

So are we sure we want to break immediately right when the first identification key is not found? Should it just be a `continue` on line 149?

Reply:
I think if `identificationKeyValue` is not null then we should add it to `identificationKeyValues` and use all existing keys for calculating hash.
```
for (final String identificationKey : identificationKeys) {
    Object identificationKeyValue = event.get(identificationKey, Object.class);
    if (identificationKeyValue != null) {
        identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
    } else {
    // Add new metric here
        recordsWithMissingIdentificationKeys.increment();
    }
 }
```

Reply:
I am not sure if it makes sense to use null value for the missing identification key. Because eventually in the aggregation the keys may be used (for example, if I am doing histogram action) and we will need those values. So, I am thinking it is better to not pass the events with missing identification keys.

Reply:
I agree with @asifsmohammed and @graytaylor0 . We should compute the hash when an identification key is missing, but others are present.

Only when no identification keys are present should we not pass these values.

Reply:
Though, I think the metric should only trigger if all values are not present.

Reply:
@dlvenable , we should use "null" for missing identification keys while calculating the hash?

Reply:
Yes, we should use `null` for calculating the hash. 

This looks compatible with the current implementation for computing the hash:

https://github.com/opensearch-project/data-prepper/blob/4196a16e38123fc36651093773b8c3511ffebe63/data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/HashRing.java#L57

Reply:
> @kkondaka , I'm still not sure how this fixes #2123. Can you share the stack trace?

ava.lang.NullPointerException: Cannot invoke "Object.toString()" because "identificationKeyValue" is null
        at org.opensearch.dataprepper.peerforwarder.RemotePeerForwarder.groupRecordsBasedOnIdentificationKeys(RemotePeerForwarder.java:156) ~[data-prepper-core-2.1.0-SNAPSHOT.jar:?]
        at org.opensearch.dataprepper.peerforwarder.RemotePeerForwarder.forwardRecords(RemotePeerForwarder.java:86) ~[data-prepper-core-2.1.0-SNAPSHOT.jar:?]
        at org.opensearch.dataprepper.peerforwarder.PeerForwardingProcessorDecorator.execute(PeerForwardingProcessorDecorator.java:81) ~[data-prepper-core-2.1.0-SNAPSHOT.jar:?]
        at org.opensearch.dataprepper.pipeline.ProcessWorker.doRun(ProcessWorker.java:95) ~[data-prepper-core-2.1.0-SNAPSHOT.jar:?]
        at org.opensearch.dataprepper.pipeline.ProcessWorker.run(ProcessWorker.java:45) ~[data-prepper-core-2.1.0-SNAPSHOT.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]

Reply:
Yes, this looks like it fixes the NPE.